### ISSUE TITLE
Question about Within-Class Online Clustering

### ISSUE THREAD

**Issue by clgx00** [tfzhou/ProtoSeg#8]
Hi, I'm interested in your work. After reading the paper, I'm confused that the goal of Within-Class Online Clustering is to map the pixels Ic to the K prototypes of class c. But how to know if pixels Ic belongs to class c? Did you use Ground Truth in this step? So how do you set it up when testing?

Hope to receive your reply, thanks!

**Comment by wenguanwang** [comment]
@clgx00 Within-class clustering is unsupervised -- for each class c, we need to automatically find K prototypes. However, the whole task setting is supervised -- for each pixel, we know the corresponding class. 

This is a compelling feature of our algorithm --  integrating unsupervised, within-class subpattern mining into supervised, per-pixel classification.

**Comment by clgx00** [comment]
@wenguanwang Yes, this is an innovative point of the algorithm. Does this mean that online clustering was not used when testing?

**Comment by wenguanwang** [comment]
@clgx00 
Of course not used. The clustering is to find class prototypes as references for classification. Once trained, the class prototypes are stored and used for classification.

**Comment by clgx00** [comment]
Thank you for your response! It is really a great work.

**Comment by yunpengt** [comment]
i want to know which part is online clustering? i can't find it..

**Comment by tfzhou** [comment]
@yunpengt please check here: https://github.com/tfzhou/ProtoSeg/blob/1c4a7784bbce96c06fe72d55255af15e6cf1ca96/lib/models/nets/hrnet.py#L54

**Comment by Fly-Yun-Maker** [comment]
Hello, thank you for your excellent work, I would like to ask what are the two values returned in the code by sink_horn, what do q and indexes mean respectively?

**Comment by QChhh123** [comment]
> 

I have the same question, do you have any idea now? Thank you very much.

**Comment by QChhh123** [comment]
> Hello, thank you for your excellent work, I would like to ask what are the two values returned in the code by sink_horn, what do q and indexes mean respectively?

I have the same question, do you have any idea now? Thank you very much.
